source/extensions/quic_listeners/quiche/active_quic_listener.cc

```diff
@@ -194,32 +260,32 @@ ActiveQuicListenerFactory::createActiveUdpListener(Network::ConnectionHandler& p
   sock_fprog prog;
   // This option only needs to be applied once to any one of the sockets in SO_REUSEPORT socket
   // group. One of the listener will be created with this socket option.
-  absl::call_once(install_bpf_once_, [&]() {
-    if (concurrency_ > 1) {
-      prog.len = filter.size();
-      prog.filter = filter.data();
-      options->push_back(std::make_shared<Network::SocketOptionImpl>(
-          envoy::config::core::v3::SocketOption::STATE_BOUND, ENVOY_ATTACH_REUSEPORT_CBPF,
-          absl::string_view(reinterpret_cast<char*>(&prog), sizeof(prog))));
-    }
-  });
-#else
-  if (concurrency_ > 1) {
-#ifdef __APPLE__
-    // Not support multiple listeners in Mac OS unless someone cares. This is because SO_REUSEPORT
-    // doesn't behave as expected in Mac OS.(#8794)
-    ENVOY_LOG(error, "Because SO_REUSEPORT doesn't guarantee stable hashing from network 5 tuple "
-                     "to socket in Mac OS. QUIC connection is not stable with concurrency > 1");
+  if (Runtime::runtimeFeatureEnabled(
+          "envoy.reloadable_features.prefer_quic_kernel_bpf_packet_routing")) {
+    absl::call_once(install_bpf_once_, [&]() {
+      if (concurrency_ > 1) {
+        prog.len = filter.size();
+        prog.filter = filter.data();
+        options->push_back(std::make_shared<Network::SocketOptionImpl>(
+            envoy::config::core::v3::SocketOption::STATE_BOUND, ENVOY_ATTACH_REUSEPORT_CBPF,
+            absl::string_view(reinterpret_cast<char*>(&prog), sizeof(prog))));
+      }
+    });
+
+    kernel_worker_routing = true;
+  };
+
 #else
-    ENVOY_LOG(warn, "BPF filter is not supported on this platform. QUIC won't support connection "
-                    "migration and NAT port rebinding.");
-#endif
+  if (concurrency_ != 1) {
+    ENVOY_LOG(
+        warn,
+        "Kernel BPF filter is not supported on this platform. QUIC performance may be degraded.");
```

Comment:
I would like to single out Mac OS because it does support BPF, but we don't have enough knowledge about its behavior.

Reply:
Right now, this log also prints on linux if the socket option isn't available. Although now that I think about it, we should probably do a runtime check to see if the socket option is available, not a compile-time check.

Also, assuming we can ever make this compile on windows, it will be covered as well.